File: src/main/java/leetcode/PalindromeLinkedList.java
package leetcode;

import datatype.ListNode;
import java.util.ArrayDeque;
import java.util.Deque;

public class PalindromeLinkedList {
  public boolean isPalindrome(ListNode head) {
    Deque<Integer> stack = new ArrayDeque<>();
    ListNode node = head;
    while (node != null) {
      stack.push(node.val);
      node = node.next;
    }
    node = head;
    while (node != null) {
      if (node.val != stack.pop()) {
        return false;
      }
      node = node.next;
    }
    return true;
  }
}

File: leetcode-kt/src/main/kotlin/com/giwankim/leetcode/PalindromeLinkedList.kt
package com.giwankim.leetcode

import java.util.Deque
import java.util.LinkedList

class PalindromeLinkedList {
    fun isPalindrome(head: ListNode?): Boolean {
        val deque: Deque<Int> = LinkedList()
        var node = head
        while (node != null) {
            deque.add(node.`val`)
            node = node.next
        }
        while (deque.size > 1) {
            if (deque.pollFirst() != deque.pollLast()) {
                return false
            }
        }
        return true
    }
}
